test/unit/reductions/exponentiated_gradient/test_lagrangian.py

```diff
@@ -0,0 +1,107 @@
+from fairlearn.reductions import DemographicParity, EqualizedOdds
+
+from .test_utilities import _get_data
+from .simple_learners import LeastSquaresBinaryClassifierLearner
+
+
+@pytest.mark.parametrize("eps", [0.001, 0.01, 0.1])
+@pytest.mark.parametrize("Constraints", [DemographicParity, EqualizedOdds])
+@pytest.mark.parametrize("h_vec", [pd.Series([1.0])])
+@pytest.mark.parametrize("opt_lambda", [True, False])
+def test_lagrangian_eval(eps, Constraints, h_vec, opt_lambda):
+    X, y, A = _get_data(A_two_dim=False)
+    estimator = LeastSquaresBinaryClassifierLearner()
+    constraints = Constraints()
+
+    # epsilon (and thereby also B) only affects L_high and L
+    B = 1 / eps
+
+    lagrangian = _Lagrangian(X, A, y, estimator, constraints, eps, B, opt_lambda=opt_lambda)
+
+    # set up initial lambda vector based on a 0-initialized theta
+    theta = pd.Series(0, lagrangian.constraints.index)
+    lambda_vec = np.exp(theta) / (1 + np.exp(theta).sum())
+
+    # call oracle to determine error and gamma and calculate exp
+    fitted_estimator = lagrangian._call_oracle(lambda_vec)
+    def h(X): return fitted_estimator.predict(X)
+    best_h_error = lagrangian.obj.gamma(h)[0]
+    best_h_gamma = lagrangian.constraints.gamma(h)
+
+    # opt_lambda affects only the calculation of L
+    if opt_lambda:
+        projected_lambda = lagrangian.constraints.project_lambda(lambda_vec)
+        L_expected = best_h_error + np.sum(projected_lambda * best_h_gamma) - \
+            eps * np.sum(projected_lambda)
+    else:
+        L_expected = best_h_error + np.sum(lambda_vec * best_h_gamma) - eps * np.sum(lambda_vec)
+
+    L_high_expected = best_h_error + B * (best_h_gamma.max() - eps)
+
+    # manually set errors and gammas which would otherwise be done in the best_h step
+    lagrangian.errors = pd.Series([best_h_error])
+    lagrangian.gammas = pd.Series([best_h_gamma])
+
+    # call _eval to get the desired results L, L_high, gamma, error
+    L, L_high, gamma, error = lagrangian._eval(h_vec, lambda_vec)
+
+    assert L == L_expected
+    assert L_high == L_high_expected
+    assert error == 0.25
+    assert (gamma == best_h_gamma).all()
+
+    # in this particular example the estimator is always the same
+    expected_estimator_weights = pd.Series({
+        'X1': 0.538136,
+        'X2': 0.457627,
+        'X3': 0.021186})
+    assert (np.isclose(fitted_estimator.weights, expected_estimator_weights, atol=1.e-6)).all()
+
+
+@pytest.mark.parametrize("Constraints", [DemographicParity, EqualizedOdds])
+@pytest.mark.parametrize("eps", [0.001, 0.01, 0.1])
+def test_call_oracle(Constraints, eps, mocker):
+    X, y, A = _get_data(A_two_dim=False)
+    estimator = mocker.MagicMock()
+    constraints = Constraints()
+
+    # ExponentiatedGradient pickles and unpickles the estimator, which isn't possible for the mock
+    # object, so we mock that process as well.
+    import pickle
```

Comment:
Shouldn't this be at the top of the file?

Reply:
I don't actually want to do anything with pickle here, so it's there to stress that it's just for this. I guess we can move it anywhere.

Reply:
I'm sure I've seen linters scream at imports anywhere than the top of the file.... although `flake8` does not appear to be complaining for now.

Reply:
Actually I wonder why it didn't complain... anyway it's changed now